source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -17,14 +21,37 @@ void ProcessorState::startMessageTimer(Event::TimerCb cb, std::chrono::milliseco
   message_timer_->enableTimer(timeout);
 }
 
-bool ProcessorState::handleHeadersResponse(const HeadersResponse& response) {
+bool ProcessorState::handleHeadersResponse(const HeadersResponse& response,
+                                           ProcessingMode_BodySendMode body_mode) {
   if (callback_state_ == CallbackState::Headers) {
     ENVOY_LOG(debug, "applying headers response");
     MutationUtils::applyCommonHeaderResponse(response, *headers_);
-    headers_ = nullptr;
     callback_state_ = CallbackState::Idle;
+    clearWatermark();
     message_timer_->disableTimer();
-    continueProcessing();
+
+    if (body_mode == ProcessingMode::BUFFERED) {
+      if (body_send_deferred_) {
+        // If we get here, then all the body data came in before the header message
+        // was complete, and the server wants the body. So, don't continue filter
+        // processing, but send the buffered request body now.
+        ENVOY_LOG(debug, "Sending buffered request body message");
+        callback_state_ = CallbackState::BufferedBody;
+        startMessageTimer(std::bind(&Filter::onMessageTimeout, parent_),
+                          parent_->config()->messageTimeout());
+        parent_->sendBodyChunk(*this, *bufferedData(), true);
+      }
+
+      // Otherwise, we're not ready to continue processing because then
+      // we won't be able to modify the headers any more, so do nothing and
+      // let the doData callback handle body chunks until the end is reached.
+
```

Comment:
Maybe an explicit return would be nice here? This comment seems to be hanging in thin air somewhat, might make it easier to correlate the comment with control flow

Reply:
I changed the return.